File: flutter/android/app/src/main/kotlin/com/dji/rc_to_controller/RcState.kt
package com.dji.rc_to_controller

data class RcState(
    val seq: Int = 0,
    val stickRightH: Int = 0,
    val stickRightV: Int = 0,
    val stickLeftV: Int = 0,
    val stickLeftH: Int = 0,
    val leftWheel: Int = 0,
    val rightWheel: Int = 0,
    val record: Boolean = false,
    val shutter: Boolean = false,
    val fiveDUp: Boolean = false,
    val fiveDDown: Boolean = false,
    val fiveDLeft: Boolean = false,
    val fiveDRight: Boolean = false,
    val fiveDCenter: Boolean = false
) {
    fun toMap(): Map<String, Any> = mapOf(
        "seq" to seq,
        "stickRightH" to stickRightH,
        "stickRightV" to stickRightV,
        "stickLeftV" to stickLeftV,
        "stickLeftH" to stickLeftH,
        "leftWheel" to leftWheel,
        "rightWheel" to rightWheel,
        "record" to record,
        "shutter" to shutter,
        "fiveDUp" to fiveDUp,
        "fiveDDown" to fiveDDown,
        "fiveDLeft" to fiveDLeft,
        "fiveDRight" to fiveDRight,
        "fiveDCenter" to fiveDCenter
    )

    companion object {
        /** Parse 18-byte HID packet into RcState. Returns null if packet is invalid. */
        fun fromHidPacket(data: ByteArray): RcState? {
            if (data.size < 18) return null
            if (data[0] != 0x02.toByte() || data[1] != 0x0E.toByte()) return null

            fun uint16LE(offset: Int): Int =
                (data[offset].toInt() and 0xFF) or ((data[offset + 1].toInt() and 0xFF) shl 8)

            val seq = uint16LE(2)
            val leftH = uint16LE(4) - 1024
            val leftV = uint16LE(6) - 1024
            val rightH = uint16LE(8) - 1024
            val rightV = uint16LE(10) - 1024
            val lWheel = uint16LE(12) - 1024
            val rWheel = uint16LE(14) - 1024

            val b16 = data[16].toInt() and 0xFF
            val b17 = data[17].toInt() and 0xFF

            return RcState(
                seq = seq,
                stickRightH = rightH,
                stickRightV = rightV,
                stickLeftV = leftV,
                stickLeftH = leftH,
                leftWheel = lWheel,
                rightWheel = rWheel,
                record = (b16 and 0x04) != 0,
                shutter = (b16 and 0x08) != 0,
                fiveDUp = (b17 and 0x01) != 0,
                fiveDDown = (b17 and 0x02) != 0,
                fiveDLeft = (b17 and 0x04) != 0,
                fiveDRight = (b17 and 0x08) != 0,
                fiveDCenter = (b17 and 0x10) != 0
            )
        }
    }
}

File: flutter/lib/models/rc_state.dart
class RcState {
  final int seq;
  final int stickLeftH;
  final int stickLeftV;
  final int stickRightH;
  final int stickRightV;
  final int leftWheel;
  final int rightWheel;
  final bool record;
  final bool shutter;
  final bool fiveDUp;
  final bool fiveDDown;
  final bool fiveDLeft;
  final bool fiveDRight;
  final bool fiveDCenter;
  final int picoBitmask;

  const RcState({
    this.seq = 0,
    this.stickLeftH = 0,
    this.stickLeftV = 0,
    this.stickRightH = 0,
    this.stickRightV = 0,
    this.leftWheel = 0,
    this.rightWheel = 0,
    this.record = false,
    this.shutter = false,
    this.fiveDUp = false,
    this.fiveDDown = false,
    this.fiveDLeft = false,
    this.fiveDRight = false,
    this.fiveDCenter = false,
    this.picoBitmask = 0,
  });

  factory RcState.fromMap(Map<dynamic, dynamic> map) {
    return RcState(
      seq: map['seq'] as int? ?? 0,
      stickLeftH: map['stickLeftH'] as int? ?? 0,
      stickLeftV: map['stickLeftV'] as int? ?? 0,
      stickRightH: map['stickRightH'] as int? ?? 0,
      stickRightV: map['stickRightV'] as int? ?? 0,
      leftWheel: map['leftWheel'] as int? ?? 0,
      rightWheel: map['rightWheel'] as int? ?? 0,
      record: map['record'] as bool? ?? false,
      shutter: map['shutter'] as bool? ?? false,
      fiveDUp: map['fiveDUp'] as bool? ?? false,
      fiveDDown: map['fiveDDown'] as bool? ?? false,
      fiveDLeft: map['fiveDLeft'] as bool? ?? false,
      fiveDRight: map['fiveDRight'] as bool? ?? false,
      fiveDCenter: map['fiveDCenter'] as bool? ?? false,
    );
  }

  Map<String, dynamic> toJson() => {
        'type': 'rc_state',
        'seq': seq,
        'stickLeftH': stickLeftH,
        'stickLeftV': stickLeftV,
        'stickRightH': stickRightH,
        'stickRightV': stickRightV,
        'leftWheel': leftWheel,
        'rightWheel': rightWheel,
        'record': record,
        'shutter': shutter,
        'fiveDUp': fiveDUp,
        'fiveDDown': fiveDDown,
        'fiveDLeft': fiveDLeft,
        'fiveDRight': fiveDRight,
        'fiveDCenter': fiveDCenter,
        'picoBitmask': picoBitmask,
      };

  RcState copyWith({int? picoBitmask}) {
    return RcState(
      seq: seq,
      stickLeftH: stickLeftH,
      stickLeftV: stickLeftV,
      stickRightH: stickRightH,
      stickRightV: stickRightV,
      leftWheel: leftWheel,
      rightWheel: rightWheel,
      record: record,
      shutter: shutter,
      fiveDUp: fiveDUp,
      fiveDDown: fiveDDown,
      fiveDLeft: fiveDLeft,
      fiveDRight: fiveDRight,
      fiveDCenter: fiveDCenter,
      picoBitmask: picoBitmask ?? this.picoBitmask,
    );
  }
}
